### PR TITLE
Fraudio interface

### PR DESCRIPTION
**Pull request by Magnus93**
Add fraudio to AuthorizationState, FailedAuthorization, PreAuthorization and model.Authorization

### CODE REVIEW

File: Fraudio/index.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	threshold: number
+	model: string
+	notes?: string
+	transaction?: string
+}
+
+export namespace Fraudio {
+	export function is(value: Fraudio | any): value is Fraudio {
+		return (
+			typeof value == "object" &&
```
Comment (robinnedergard) [inline Fraudio/index.ts:17]: Implicit not null check:
```suggestion
			typeof value == "object" && value &&
```

File: Fraudio/index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import * as selectively from "selectively"
+import * as authly from "authly"
+
+export interface Fraudio {
+	score: number
+	recommendation: "green" | "yellow" | "red"
+	id: authly.Identifier
```
Comment (robinnedergard) [inline Fraudio/index.ts:7]: string, never assume externally provided id:s are dividable by 4 (which authly.Identifier always are)
```suggestion
	id: string
```

File: Fraudio/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	transaction?: string
+}
+
+export namespace Fraudio {
+	export function is(value: Fraudio | any): value is Fraudio {
+		return (
+			typeof value == "object" &&
+			typeof value.score == "number" &&
+			["green", "yellow", "red"].includes(value.recommendation) &&
+			authly.Identifier.is(value.id) &&
```
Comment (robinnedergard) [inline Fraudio/index.ts:20]:
```suggestion
			typeof value.id == "string" &&
```

File: Fraudio/index.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import * as selectively from "selectively"
+import * as authly from "authly"
+
+export interface Fraudio {
+	score: number
```
Comment (robinnedergard) [inline Fraudio/index.ts:5]: Original value is actually a string containing a number
```suggestion
	score: string
```

File: Fraudio/index.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	recommendation: "green" | "yellow" | "red"
+	id: authly.Identifier
+	threshold: number
```
Comment (robinnedergard) [inline Fraudio/index.ts:8]: Original value is actually a string containing a number
```suggestion
	threshold: string
```

File: Fraudio/index.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+export namespace Fraudio {
+	export function is(value: Fraudio | any): value is Fraudio {
+		return (
+			typeof value == "object" &&
+			typeof value.score == "number" &&
+			["green", "yellow", "red"].includes(value.recommendation) &&
+			authly.Identifier.is(value.id) &&
+			typeof value.threshold == "number" &&
```
Comment (robinnedergard) [inline Fraudio/index.ts:21]:
```suggestion
			typeof value.threshold == "string" &&
```

File: Fraudio/index.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	model: string
+	notes?: string
+	transaction?: string
+}
+
+export namespace Fraudio {
+	export function is(value: Fraudio | any): value is Fraudio {
+		return (
+			typeof value == "object" &&
+			typeof value.score == "number" &&
```
Comment (robinnedergard) [inline Fraudio/index.ts:18]:
```suggestion
			typeof value.score == "string" &&
```

File: Fraudio/index.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+			id: value.Fraudio_Transaction_ID,
+			threshold: value.Threshold,
+			model: value.Model,
+			notes: value.Notes,
+			transaction: value.Transaction,
+		}
+		return is(result) ? result : undefined
+	}
+	export const template = new selectively.Type.Object({
+		score: new selectively.Type.Number(),
```
Comment (robinnedergard) [inline Fraudio/index.ts:40]:
```suggestion
		score: new selectively.Type.String(),
```

File: Fraudio/index.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+		recommendation: new selectively.Type.Union(["green", "yellow", "red"].map(r => new selectively.Type.String(r))),
+		id: new selectively.Type.String(),
+		threshold: new selectively.Type.Number(),
```
Comment (robinnedergard) [inline Fraudio/index.ts:43]:
```suggestion
		threshold: new selectively.Type.String(),
```

File: State/Authorization.spec.ts [hunk 10]
```diff
@@ -46,6 +46,13 @@ describe("State.Authorization", () => {
 			history: [],
 			created: "2021-01-01T12:30:30.000Z",
 			category: "withdrawal",
+			fraudio: {
+				score: 0.75,
+				recommendation: "yellow",
+				id: "12345678",
+				threshold: 0.5,
+				model: "modelY",
```
Comment (robinnedergard) [inline State/Authorization.spec.ts:54]:
```suggestion
				score: "0.75",
				recommendation: "yellow",
				id: "12345678",
				threshold: "0.5",
				model: "modelY",
```

**Review by robinnedergard — CHANGES_REQUESTED** [review]
Also add parity

File: Fraudio/index.ts (hunk 4 above)
Comment (Magnus93) [reply Fraudio/index.ts:5]: I'll convert to number in the from function.

File: Fraudio/index.ts (hunk 5 above)
Comment (Magnus93) [reply Fraudio/index.ts:8]: I'll convert to number in the from function

File: Fraudio/index.ts (hunk 1 above)
Comment (Magnus93) [reply Fraudio/index.ts:17]: Fixed

File: Authorization/index.ts [hunk 11]
```diff
@@ -121,6 +124,8 @@ export namespace Authorization {
 									property: "category",
 									type: `"purchase" | "withdrawal" | undefined`,
 								},
+							value.category == undefined ||
```
Comment (Magnus93) [inline Authorization/index.ts:127]:
```suggestion
							value.fraudio == undefined ||
```

**Review by robinnedergard — APPROVED** [review]

**Review by nilssonemma — APPROVED** [review]

**Review by marcushanses — APPROVED** [review]

**Review by fredrikbryntesson — APPROVED** [review]

**Pull request merged**